src/frequenz/channels/util/_select.py

```diff
@@ -11,13 +11,598 @@
+
+    This exception is raised when a `Selector` loop iteration fails.  It is raised as
+    a single exception when one receiver fails during normal operation (while calling
+    `ready()` for example).  It is raised as a group exception
+    ([`SelectErrorGroup`][frequenz.channels.util.SelectErrorGroup]) when a `Selector` is
+    [`stop()`][frequenz.channels.util.Selector.stop]ed.
+    """
+
+
+class UnhandledSelectedError(SelectError, Generic[_T]):
+    """A receiver was not handled in a [`Selector`][frequenz.channels.util.Selector] loop.
+
+    This exception is raised when a `Selector` loop finishes without a call to
+    [`selected_from()`][frequenz.channels.util.selected_from] for the selected receiver.
+    """
+
+    def __init__(self, selected: Selected[_T]) -> None:
+        """Create a new instance.
+
+        Args:
+            selected: The selected receiver that was not handled.
+        """
+        recv = selected._recv  # pylint: disable=protected-access
+        super().__init__(f"Selected receiver {recv} was not handled in the if-chain")
+        self.selected = selected
+
+
+class SelectErrorGroup(BaseExceptionGroup[BaseException], SelectError):
+    """An exception group for [`Selector`][frequenz.channels.util.Selector].
+
+    This exception group is raised when a [`Selector.stop()`] fails while cleaning up
+    runing tasts to check for ready receivers.
+    """
+
+
+# Typing for select() is tricky.  We had the idea of using a declarative design for
+# select, something like:
+#
+# ```python
+# class MySelector(Selector):
+#     receiver1: x.new_receiver()
+#     receiver2: y.new_receiver()
+#
+# async for selected in MySelector:
+#     if selected.receiver is receiver1:
+#         # Do something with selected.value
+#     elif selected.receiver is receiver1:
+#         # Do something with selected.value
+# ```
+#
+# This is similar to `Enum`, but `Enum` has special support in `mypy` that we can't
+# have.
+#
+# With the current implementation, the typing could be slightly improved by using
+# `TypeVarTuple`, but we are not because "transformations" are not supported yet, see:
+# https://github.com/python/typing/issues/1216
+#
+# Also support for `TypeVarTuple` in general is still experimental (and very incomplete
+# in `mypy`).
+#
+# With this we would also probably be able to properly type `select` and *maybe* even be
+# able to leverage the exhaustiveness checking of `mypy` to make sure the selected value
+# is narrowed down to the correct type to make sure all receivers are handled, with the
+# help of `assert_never` as described in:
+# https://docs.python.org/3.11/library/typing.html#typing.assert_never
+#
+# We also explored the possibility of using `match` to perform exhaustiveness checking,
+# but we couldn't find a way to make it work with `match`, and `match` is not yet
+# checked for exhaustiveness by `mypy` anyway, see:
+# https://github.com/python/mypy/issues/13597
+
+
+class Selector:
```

Comment:
I think this should be renamed to `Select` because it is not used like a regular Python class which we create and call methods on.  It is used like a select function,  and is implemented as a class because of Python language limitations.  And it would then also be the same idiom from other languages.

Reply:
I named it `Selector` because for me it is still mainly an async generator, not really a function, there is even no function call syntax involved to use it (except for creating the object). Even with the previous implementation it was weird to me to call it `Select`, because if was even more an object than this implementation, as you did have to call methods (like `ready()` on it).

Naming discussions are complicated and at the end I think the goal is to use whatever feels more obvious to users, so I will create a poll in our internal chat to see what people think is more readable/obvious, hope that is OK with you.

Reply:
`select()` is back! The name discussion is moot now :tada: